File: evaluation/S0/c/Medium/1356-can-you-eat-your-favorite-candy-on-your-favorite-day.c
// Source: https://leetcode.com/problems/can-you-eat-your-favorite-candy-on-your-favorite-day/   |   Difficulty: Medium
//
// Problem Description:
// You are given a (0-indexed) array of positive integers candiesCount where candiesCount[i] represents the number of candies of the ith type you have. You are also given a 2D array queries where queries[i] = [favoriteTypei, favoriteDayi, dailyCapi].
//
// You play a game with the following rules:
//
//
// 	You start eating candies on day 0.
// 	You cannot eat any candy of type i unless you have eaten all candies of type i - 1.
// 	You must eat at least one candy per day until you have eaten all the candies.
//
//
// Construct a boolean array answer such that answer.length == queries.length and answer[i] is true if you can eat a candy of type favoriteTypei on day favoriteDayi without eating more than dailyCapi candies on any day, and false otherwise. Note that you can eat different types of candy on the same day, provided that you follow rule 2.
//
// Return the constructed array answer.
//
// Example:
// Input: candiesCount = [7,4,5,3,8], queries = [[0,2,2],[4,2,4],[2,13,1000000000]]
// Output: [true,false,true]
// Explanation:
// 1- If you eat 2 candies (type 0) on day 0 and 2 candies (type 0) on day 1, you will eat a candy of type 0 on day 2.
// 2- You can eat at most 4 candies each day.
//    If you eat 4 candies every day, you will eat 4 candies (type 0) on day 0 and 4 candies (type 0 and type 1) on day 1.
//    On day 2, you can only eat 4 candies (type 1 and type 2), so you cannot eat a candy of type 4 on day 2.
// 3- If you eat 1 candy each day, you will eat a candy of type 2 on day 13.
//
// Constraints:
// 1 <= candiesCount.length <= 105
// 	1 <= candiesCount[i] <= 105
// 	1 <= queries.length <= 105
// 	queries[i].length == 3
// 	0 <= favoriteTypei < candiesCount.length
// 	0 <= favoriteDayi <= 109
// 	1 <= dailyCapi <= 109
//

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
bool* canEat(int* candiesCount, int candiesCountSize, int** queries, int queriesSize, int* queriesColSize, int* returnSize) {
    // Precompute prefix sums of candiesCount
    long long* prefixSums = (long long*)malloc((candiesCountSize + 1) * sizeof(long long));
    prefixSums[0] = 0;
    for (int i = 0; i < candiesCountSize; i++) {
        prefixSums[i + 1] = prefixSums[i] + candiesCount[i];
    }

    // Prepare the result array
    bool* result = (bool*)malloc(queriesSize * sizeof(bool));
    *returnSize = queriesSize;

    // Process each query
    for (int i = 0; i < queriesSize; i++) {
        int favoriteType = queries[i][0];
        long long favoriteDay = queries[i][1];
        long long dailyCap = queries[i][2];

        // Calculate the minimum and maximum candies that can be eaten by favoriteDay
        long long minCandiesByDay = favoriteDay + 1; // At least one candy per day
        long long maxCandiesByDay = (favoriteDay + 1) * dailyCap; // At most dailyCap candies per day

        // Calculate the range of candies for the favorite type
        long long candiesBeforeType = prefixSums[favoriteType];
        long long candiesUpToType = prefixSums[favoriteType + 1];

        // Check if there is an overlap between [minCandiesByDay, maxCandiesByDay] and [candiesBeforeType + 1, candiesUpToType]
        if (maxCandiesByDay > candiesBeforeType && minCandiesByDay <= candiesUpToType) {
            result[i] = true;
        } else {
            result[i] = false;
        }
    }

    free(prefixSums);
    return result;      
}

File: evaluation/S3/python/Medium/1356-can-you-eat-your-favorite-candy-on-your-favorite-day.py
# Source: https://leetcode.com/problems/can-you-eat-your-favorite-candy-on-your-favorite-day/   |   Difficulty: Medium
#
# Problem Description:
# You are given a (0-indexed) array of positive integers candiesCount where candiesCount[i] represents the number of candies of the ith type you have. You are also given a 2D array queries where queries[i] = [favoriteTypei, favoriteDayi, dailyCapi].
#
# You play a game with the following rules:
#
#
# 	You start eating candies on day 0.
# 	You cannot eat any candy of type i unless you have eaten all candies of type i - 1.
# 	You must eat at least one candy per day until you have eaten all the candies.
#
#
# Construct a boolean array answer such that answer.length == queries.length and answer[i] is true if you can eat a candy of type favoriteTypei on day favoriteDayi without eating more than dailyCapi candies on any day, and false otherwise. Note that you can eat different types of candy on the same day, provided that you follow rule 2.
#
# Return the constructed array answer.
#
# Example:
# Input: candiesCount = [7,4,5,3,8], queries = [[0,2,2],[4,2,4],[2,13,1000000000]]
# Output: [true,false,true]
# Explanation:
# 1- If you eat 2 candies (type 0) on day 0 and 2 candies (type 0) on day 1, you will eat a candy of type 0 on day 2.
# 2- You can eat at most 4 candies each day.
#    If you eat 4 candies every day, you will eat 4 candies (type 0) on day 0 and 4 candies (type 0 and type 1) on day 1.
#    On day 2, you can only eat 4 candies (type 1 and type 2), so you cannot eat a candy of type 4 on day 2.
# 3- If you eat 1 candy each day, you will eat a candy of type 2 on day 13.
#
# Constraints:
# 1 <= candiesCount.length <= 105
# 	1 <= candiesCount[i] <= 105
# 	1 <= queries.length <= 105
# 	queries[i].length == 3
# 	0 <= favoriteTypei < candiesCount.length
# 	0 <= favoriteDayi <= 109
# 	1 <= dailyCapi <= 109
#
# Helpful references (internal KB):
# - Fenwick Tree (array, bit-fenwick, prefix-sum)
#   • When to use: When you need to efficiently calculate prefix sums or range sums on an array that undergoes frequent point updates. It is ideal for scenarios requiring both fast query and update operations on an array.
#   • Idea: A Fenwick tree, or Binary Indexed Tree (BIT), is an array-based data structure that efficiently calculates prefix sums and updates individual elements in an array. It supports both point updates and prefix sum queries in O(log N) time complexity.
#   • Invariants: Each `T[i]` stores the sum of a specific range `[g(i), i]` from the original array.; The `sum(idx)` function correctly aggregates values from `A[0]` to `A[idx]`.
#   • Tips: Understand the bit manipulation (`i & -i`) for navigating parent/child relationships.; Consistently choose and apply either 0-based or 1-based indexing throughout the implementation.
#   • Pitfalls: Incorrectly handling 0-based versus 1-based indexing can lead to off-by-one errors.; Forgetting to initialize the tree or performing updates incorrectly will yield wrong results.
# - Sqrt Decomposition (array, enumeration, difference-array)
#   • When to use: Efficiently calculate sums or properties over various ranges in a static array. Useful for many range queries or when combining range updates with point queries.
#   • Idea: Prefix sums allow O(1) range sum queries after O(N) preprocessing. Difference arrays enable O(1) range updates and O(N) point queries after O(N) reconstruction.
#   • Invariants: The prefix sum array P[i] stores the sum of elements from A[0] to A[i-1].; The difference array D[i] stores A[i] - A[i-1] (for i > 0).
#   • Tips: Precompute prefix sums to answer range sum queries in constant time.; Use a difference array to apply range updates efficiently.
#   • Pitfalls: Off-by-one errors in range boundaries for prefix sums.; Forgetting to initialize the difference array correctly (e.g., diff[0] = arr[0] for 0-indexed).
# - Disjoint Sparse Table tutorial (array, ds, number, sparse-table, counting)
#   • When to use: Use when you need to answer range queries on a static array in O(1) time, after an initial O(N log N) preprocessing phase. It is particularly efficient for associative operations like sum, min, or max.
#   • Idea: A Disjoint Sparse Table is a data structure that precomputes results for all possible range queries on a static array using a hierarchical block decomposition. It achieves O(N log N) construction time and space complexity, enabling O(1) query time for associative operations.
#   • Invariants: ST[k][i] stores the result for a range of length 2^k starting at i within its block.; For any level k, blocks are disjoint, but ranges within ST[k] can overlap with ranges in ST[k-1].
#   • Tips: Precompute log2 values or highest set bit for efficient level determination.; Ensure the operation (e.g., sum, min, max) is associative for correct combination.
#   • Pitfalls: Not suitable for arrays that require updates, as it's a static data structure.; Only works for associative operations; non-associative operations cannot be combined this way.
#
# Notes:
# - Keep the given function signature unchanged.
# - Use only the Python standard library; no third-party packages.
# - Do not change input/output formats or return value semantics as defined by the template/problem.
# - Conformance: use features compatible with LeetCode’s Python3 runtime; do NOT rely on undefined/implementation-specific behavior.
# - If you introduce any helper functions/classes, include their full implementations in this file; no hidden helpers.

class Solution(object):
    def canEat(self, candiesCount, queries):
        """
        :type candiesCount: List[int]
        :type queries: List[List[int]]
        :rtype: List[bool]
        """
        prefix_sum = [0] * len(candiesCount)
        prefix_sum[0] = candiesCount[0]
        for i in range(1, len(candiesCount)):
            prefix_sum[i] = prefix_sum[i - 1] + candiesCount[i]

        result = []
        for favoriteType, favoriteDay, dailyCap in queries:
            min_candies_needed = prefix_sum[favoriteType - 1] + 1 if favoriteType > 0 else 1
            max_candies_possible = prefix_sum[favoriteType]
            min_candies_eaten_by_day = favoriteDay + 1
            max_candies_eaten_by_day = (favoriteDay + 1) * dailyCap

            can_eat = (min_candies_needed <= max_candies_eaten_by_day) and (max_candies_possible >= min_candies_eaten_by_day)
            result.append(can_eat)

        return result